packages/service-library/src/servicelib/fastapi/long_running_tasks/_context_manager.py

```diff
@@ -72,33 +82,87 @@ async def periodic_task_result(
         the expected result
     raises: `asyncio.TimeoutError` NOTE: the remote task will also be removed
     """
+
     progress_manager = _ProgressManager(progress_callback)
 
-    async def _status_update() -> TaskStatus:
-        task_status = await client.get_task_status(task_id)
-        logger.info("Task status %s", task_status.json())
-        await progress_manager.update(
-            message=task_status.task_progress.message,
-            percent=task_status.task_progress.percent,
+    async def _statuses_update() -> list[TaskStatus]:
+        tasks_status: list[TaskStatus] = await logged_gather(
+            *(client.get_task_status(task_id) for task_id in task_ids),
+            max_concurrency=MAX_CONCURRENCY,
         )
-        return task_status
-
-    async def _wait_task_completion() -> None:
-        task_status = await _status_update()
-        while not task_status.done:
+        for task_id, task_status in zip(task_ids, tasks_status):
+            logger.debug("Task status %s", task_status.json())
+            await progress_manager.update(
+                task_id=task_id,
+                message=task_status.task_progress.message,
+                percent=task_status.task_progress.percent,
+            )
+
+        return tasks_status
+
+    async def _wait_tasks_completion() -> None:
+        tasks_statuses = await _statuses_update()
+        while not all(task_status.done for task_status in tasks_statuses):
             await asyncio.sleep(status_poll_interval)
-            task_status = await _status_update()
+            tasks_statuses = await _statuses_update()
 
     try:
-        await asyncio.wait_for(_wait_task_completion(), timeout=task_timeout)
+        await asyncio.wait_for(_wait_tasks_completion(), timeout=task_timeout)
+
+        results: list[Optional[Any]] = await logged_gather(
+            *(client.get_task_result(task_id) for task_id in task_ids),
+            max_concurrency=MAX_CONCURRENCY,
+        )
+        for result, task_id in zip(results, task_ids):
+            logger.debug("Task %s result %s", task_id, result)
 
-        result: Optional[Any] = await client.get_task_result(task_id)
-        yield result
+        yield results
     except asyncio.TimeoutError as e:
-        task_removed = await client.cancel_and_delete_task(task_id)
+        tasks_removed: list[bool] = await logged_gather(
+            *(client.cancel_and_delete_task(task_id) for task_id in task_ids),
```

Comment:
can this fail?

Reply:
Yes, I've added a retry on `httpx.HttpError`. It will get retried 3 times before giving up